File: sdk/typescript/src/connectivity-provider.ts
import type { Extension } from "./extension.js";
import type { ExtensionContext } from "./extension-contract.js";
import { registerProvideSurface } from "./extension-provide-surface.js";
import { requestRecord, requiredString } from "./protocol-params.js";
import type {
  ConnectivityEstablishRequest,
  ConnectivityReachability,
  ConnectivityStatusRequest,
  ConnectivityTeardownRequest,
  ConnectivityTeardownResponse,
} from "./types.js";

export const CONNECTIVITY_PROVIDER_CAPABILITY = "connectivity.provider";
export const CONNECTIVITY_ESTABLISH_METHOD = "connectivity/establish";
export const CONNECTIVITY_STATUS_METHOD = "connectivity/status";
export const CONNECTIVITY_TEARDOWN_METHOD = "connectivity/teardown";

export interface ConnectivityProviderHandlers {
  establish: (
    context: ExtensionContext,
    request: ConnectivityEstablishRequest
  ) => Promise<ConnectivityReachability> | ConnectivityReachability;
  status: (
    context: ExtensionContext,
    request: ConnectivityStatusRequest
  ) => Promise<ConnectivityReachability> | ConnectivityReachability;
  teardown: (
    context: ExtensionContext,
    request: ConnectivityTeardownRequest
  ) => Promise<ConnectivityTeardownResponse> | ConnectivityTeardownResponse;
}

function parseEstablishRequest(request: unknown): ConnectivityEstablishRequest {
  const record = requestRecord(CONNECTIVITY_ESTABLISH_METHOD, request);
  return {
    tier: requiredString(CONNECTIVITY_ESTABLISH_METHOD, record, "tier"),
    forward_target: requiredString(CONNECTIVITY_ESTABLISH_METHOD, record, "forward_target"),
    challenge_path: requiredString(CONNECTIVITY_ESTABLISH_METHOD, record, "challenge_path"),
    deadline: requiredString(CONNECTIVITY_ESTABLISH_METHOD, record, "deadline"),
  };
}

function parseStatusRequest(request: unknown): ConnectivityStatusRequest {
  const record = requestRecord(CONNECTIVITY_STATUS_METHOD, request);
  return {
    tier: requiredString(CONNECTIVITY_STATUS_METHOD, record, "tier"),
  };
}

function parseTeardownRequest(request: unknown): ConnectivityTeardownRequest {
  const record = requestRecord(CONNECTIVITY_TEARDOWN_METHOD, request);
  return {
    tier: requiredString(CONNECTIVITY_TEARDOWN_METHOD, record, "tier"),
    deadline: requiredString(CONNECTIVITY_TEARDOWN_METHOD, record, "deadline"),
  };
}

export function registerConnectivityProvider(
  extension: Extension,
  handlers: ConnectivityProviderHandlers
): Extension {
  if (!extension) {
    throw new Error("extension is required");
  }
  if (
    typeof handlers?.establish !== "function" ||
    typeof handlers.status !== "function" ||
    typeof handlers.teardown !== "function"
  ) {
    throw new Error("connectivity provider handlers are required");
  }
  return extension[registerProvideSurface](CONNECTIVITY_PROVIDER_CAPABILITY, [
    {
      method: CONNECTIVITY_ESTABLISH_METHOD,
      handler: async (context, request) =>
        await handlers.establish(context, parseEstablishRequest(request)),
    },
    {
      method: CONNECTIVITY_STATUS_METHOD,
      handler: async (context, request) =>
        await handlers.status(context, parseStatusRequest(request)),
    },
    {
      method: CONNECTIVITY_TEARDOWN_METHOD,
      handler: async (context, request) =>
        await handlers.teardown(context, parseTeardownRequest(request)),
    },
  ]);
}

File: sdk/go/connectivity_provider.go
package compozysdk

import (
	"context"
	"encoding/json"
	"strings"

	"github.com/compozy/compozy/sdk/go/contracts"
)

// ConnectivityEstablishRequest is the daemon request to start one connectivity tier.
type ConnectivityEstablishRequest = contracts.ConnectivityEstablishRequest

// ConnectivityStatusRequest is the daemon request to inspect one connectivity tier.
type ConnectivityStatusRequest = contracts.ConnectivityStatusRequest

// ConnectivityTeardownRequest is the daemon request to stop one connectivity tier.
type ConnectivityTeardownRequest = contracts.ConnectivityTeardownRequest

// ConnectivityReachability reports the provider's current endpoints and health.
type ConnectivityReachability = contracts.ConnectivityReachability

// ConnectivityAdvertisedEndpoint is one endpoint claimed by a connectivity provider.
type ConnectivityAdvertisedEndpoint = contracts.ConnectivityAdvertisedEndpoint

// ConnectivityTeardownResponse reports whether a connectivity tier stopped.
type ConnectivityTeardownResponse = contracts.ConnectivityTeardownResponse

// ConnectivityProviderHandlers implements the three daemon-initiated provider services.
type ConnectivityProviderHandlers struct {
	Establish func(context.Context, ExtensionContext, ConnectivityEstablishRequest) (ConnectivityReachability, error)
	Status    func(context.Context, ExtensionContext, ConnectivityStatusRequest) (ConnectivityReachability, error)
	Teardown  func(context.Context, ExtensionContext, ConnectivityTeardownRequest) (ConnectivityTeardownResponse, error)
}

// ConnectivityProvider registers one public reachability provider and advertises
// CapabilityProvideConnectivityProvider on the extension runtime.
func ConnectivityProvider(extension *Extension, handlers ConnectivityProviderHandlers) error {
	if extension == nil {
		return NewInternalError("extension is required")
	}
	if handlers.Establish == nil || handlers.Status == nil || handlers.Teardown == nil {
		return NewInvalidParamsError("connectivity provider handlers are required", nil)
	}
	registrations := []struct {
		method  string
		handler ExtensionHandler
	}{
		{ExtensionServiceMethodConnectivityEstablish, typedConnectivityHandler(handlers.Establish)},
		{ExtensionServiceMethodConnectivityStatus, typedConnectivityHandler(handlers.Status)},
		{ExtensionServiceMethodConnectivityTeardown, typedConnectivityHandler(handlers.Teardown)},
	}
	extension.mu.Lock()
	defer extension.mu.Unlock()
	if extension.initialized {
		return NewInvalidParamsError("extension registration is closed after initialize", nil)
	}
	if extension.connectivityProvider {
		return NewInvalidParamsError("connectivity provider is already registered", nil)
	}
	for _, registration := range registrations {
		if _, exists := extension.handlers[registration.method]; exists {
			return NewInvalidParamsError(registration.method+" is reserved by ConnectivityProvider", nil)
		}
	}
	for _, registration := range registrations {
		extension.handlers[registration.method] = registration.handler
		extension.transport.Handle(registration.method, extension.dispatch)
	}
	extension.connectivityProvider = true
	extension.definition.Capabilities.Provides = normalizeStrings(append(
		extension.definition.Capabilities.Provides,
		CapabilityProvideConnectivityProvider,
	))
	return nil
}

func typedConnectivityHandler[Request any, Response any](
	handler func(context.Context, ExtensionContext, Request) (Response, error),
) ExtensionHandler {
	return func(ctx context.Context, extensionContext ExtensionContext, raw json.RawMessage) (any, error) {
		var request Request
		if len(raw) == 0 {
			return nil, NewInvalidParamsError("connectivity request is required", nil)
		}
		if err := json.Unmarshal(raw, &request); err != nil {
			return nil, NewInvalidParamsError("connectivity request is invalid", map[string]any{
				"error": err.Error(),
			})
		}
		response, err := handler(ctx, extensionContext, request)
		if err != nil {
			return nil, err
		}
		return response, nil
	}
}

func isConnectivityProviderMethod(method string) bool {
	switch strings.TrimSpace(method) {
	case ExtensionServiceMethodConnectivityEstablish,
		ExtensionServiceMethodConnectivityStatus,
		ExtensionServiceMethodConnectivityTeardown:
		return true
	default:
		return false
	}
}
